prover/man/man4/invcheck.4
.TH INVCHECK 4 19/03/2001
.SH NAME
invcheck \- Check an invariant for a linear process (LPO)

.SH SYNOPSIS
.B invcheck
-invariant 
.I 
inv-file\c
 [OPTIONS]

.SH OPTIONS 
.IP \-generate
writes the invariance formulae to stdout, instead of proving them.
.IP \-all
doesn't terminate as soon as an invariant-violation is found
.IP \-verbose
indicates on stderr for which summands the invariant holds
.IP \-print
prints the resulting BDDs
.IP \-counter
prints counter-examples for summands where the invariant is not preserved
.IP \-help
prints a help message

.in 0.75i
If -generate is absent, -verbose is turned on automatically. Furthermore,
-generate cannot be used in combination with -print or -counter.

.SH DESCRIPTION
.B invcheck
is a tool that checks and reports whether the formula in 
.I inv-file
holds for the initial state, and whether it is preserved by all
transitions in the LPO on stdin.  With the -counter and -print
options, one gets some diagnostic information.

With the -generate flag, the invariance formulas are generated in the
format understood by formcheck. These can be solved by another prover.

.SH FILES
.IP inv-file
A file containing a formula, which is a term of type Bool, containing
the LPO parameters as free variables.
.IP stdin
An LPO in tbf-format.
.IP stdout
A list of declarations and formulae, only generated with option -generate.

.SH SEE ALSO

prover(4), formcheck(4)

.SH AUTHOR
Jaco van de Pol <vdpol@cwi.nl>
.LP
Please send all complaints, comments and bug fixes to me. 


svc/ChangeLog
   SVC -- the SVC (Systems Validation Centre) file format library

   Copyright (C) 2000  Stichting Mathematisch Centrum, Amsterdam,
                       The  Netherlands

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

$Id: ChangeLog,v 1.1.1.1 2004/09/07 15:06:33 uid523 Exp $



2000/6/1
   * Version 1.0 
   * Initial release

2000/6/15 (izak@cwi.nl)
   * Version 1.1 
   * Bug fix: frequency of the escape sequence is not updated

2000/9/18 (izak@cwi.nl)
   * Version 1.2
   * Update conform update ATerm library 1.4.4
   * Added configure.in to release

2000/9/25
   * Release 1.0.3
   * Fixed bug in troff manuals causing manuals to start with a blank page
   * Fixed memory allocation bug in compressed streams
   * Cleaned up header files by removing standard includes

2000/9/27
   * Release 1.0.4
   * Fixed bug: buffer for states and labels was allocated too small

2000/10/18
   * Release 1.0.5
   * Fixed bug: the SVCaterm2... functions returned 0 instead of -1 to flag
     an unknown ATerm

2001/1/4
   * Release 1.0.6
   * Format version number incremented to 1.2 beta
   * Lempel Ziff token enlarged to three bytes
   * Gnu General Public License added


prover/src/invcheck.c
/*  $Id: invcheck.c,v 1.3 2005/03/11 16:55:15 bertl Exp $
This program generates the invariance conditions for an LPO,

   12/07/2000 Jaco van de Pol
   - hold in initial state
   - are preserved for each summand
    
   8/12/2000  and also checks them.

*/
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include "aterm2.h"
#include "mcrl.h"
#include "prover.h"
#include <assert.h>

#define MY_NAME "Invcheck"
// VERSION assumed to be defined with -DVERSION=\"...\"

char PRINT=0;
char DOT=0;
char COUNTER=0;
static char VERBOSE=0;
char GENERATE=0;
static char ALL=0;

char* Inv_file=0;

void version() {
  fprintf(stderr,"%s (version %s)\n",MY_NAME, VERSION);
  exit(0);
}

void help() {
  ATfprintf(stderr,"%s\n",
	    "invcheck -invariant file1 [options] [file2]\n"
	    "-- checks whether the invariant holds");
  ATfprintf(stderr,"%s\n","");
  ATfprintf(stderr,"%s\n","Reads LPO from [file2[.tbf]] or stdin and invariant from file1");
  ATfprintf(stderr,"%s\n","");
  ATfprintf(stderr,"%s\n","OPTIONS:");
  ATfprintf(stderr,"%s\n",
	    "-generate      : "
	    "generates formulae without checking to stdout");
  ATfprintf(stderr,"%s\n",
	    "-all           : "
	    "doesn't terminate as soon as an invariant-violation is found");
  ATfprintf(stderr,"%s\n",
	    "-verbose       : "
	    "indicates for which summands the invariant holds");
  ATfprintf(stderr,"%s\n",
	    "-print         : "
	    "prints BDDs for failed summands in ASCII on stderr");
  ATfprintf(stderr,"%s\n",
	    "-print-dot     : "
	    "prints BDDs for failed summands in dot-format on stdout");
  ATfprintf(stderr,"%s\n",
	    "-counter       : "
	    "prints counter-examples for failed summands on stderr");
  ATfprintf(stderr,"%s\n",
	    "-version       : "
	    "prints the version number");
  ATfprintf(stderr,"%s\n",
	    "-help          : "
	    "prints this message");
  ATfprintf(stderr,"%s\n",
	    "-help-mcrl     : "
	    "prints standard options of mcrl toolset");

  ATfprintf(stderr,"%s\n","");
  ATfprintf(stderr,"%s\n","If -generate is absent, -verbose is on automatically");
  ATfprintf(stderr,"%s\n","Don't use '-generate' together with '-counter/print'");
  exit(0);
}

void parse_argument_line(int *argc,char **argv[]) {
  int i;
  int newargc=1;
  char** newargv=(char**)calloc(*argc,sizeof(char*));
  newargv[0]=(*argv)[0];
  
  for (i=1;i<*argc;i++) {
    if (!strcmp((*argv)[i],"-counter"))
      COUNTER=1; 
    else if (!strcmp((*argv)[i],"-all"))
      ALL=1;
    else if (!strcmp((*argv)[i],"-print-dot"))
      DOT=1; 
    else if (!strcmp((*argv)[i],"-print"))
      PRINT=1; 
    else if (!strcmp((*argv)[i],"-verbose"))
      VERBOSE=1; 
    else if (!strcmp((*argv)[i],"-generate"))
      GENERATE=1; 
    else if (!strcmp((*argv)[i],"-invariant"))
      Inv_file=(*argv)[++i]; 
    else if (!strcmp((*argv)[i],"-help"))
      help(); 
    else if (!strcmp((*argv)[i],"-help-mcrl")) {
      MCRLhelp();
      exit(0);
    }
    else if (!strcmp((*argv)[i],"-version"))
      version();
    else 
      newargv[newargc++]=(*argv)[i];
  }
  *argv=newargv;
  *argc=newargc;

  if (!Inv_file) ATerror("Invariant file required '-invariant file'\n");
  if (!GENERATE) VERBOSE=1;
  if (GENERATE && (DOT || PRINT || COUNTER))
    fprintf(stderr,"'-print/-print-dot/-counter' is ignored with '-generate'\n");
}

static ATerm substitute_const(ATermList termlist, ATermList varlist,ATerm v)
{ ATerm var;
  while (!ATisEmpty(termlist)) {
    assert(!ATisEmpty(varlist));
    var=ATgetFirst(varlist); 
    // var is of the form v(name,sort)
    if (v==ATgetArgument(var,0)) 
      return ATgetFirst(termlist);
    termlist=ATgetNext(termlist);
    varlist=ATgetNext(varlist);
  }
  assert(ATisEmpty(varlist));
  return v;
}

static ATerm substitute(ATermList termlist, ATermList varlist,ATerm t);

static ATermList substitute_list(ATermList termlist, ATermList varlist,ATermList tl) {
  ATerm t1;
  if (ATisEmpty(tl)) return tl;
  
  t1 = substitute(termlist,varlist,ATgetFirst(tl));
  tl = substitute_list(termlist,varlist,ATgetNext(tl));
  return ATinsert(tl,t1);
}

static ATerm substitute(ATermList termlist, ATermList varlist,ATerm t)
{ 
  Symbol f;
  f = ATgetSymbol(t);
  if (ATgetArity(f)==0)
    { return substitute_const(termlist,varlist,t);
    }
  return (ATerm)ATmakeApplList(f,
			       substitute_list(termlist,varlist,
					       ATgetArguments((ATermAppl)t)));
}

static int check_init(ATermList init,ATermList parameters, ATerm I) {
  ATerm formula;
  formula = substitute(init,parameters,I);

  if (GENERATE) {
    ATfprintf(stdout,"%t,\n", prettyprint_decls(parameters));
    ATfprintf(stdout,"%t" , prettyprint(Simplify(formula)));
    return 1;
  }
  
  formula = Prove(formula);
  if (formula==prTRUE) {
    if (VERBOSE) fprintf(stderr,"Init OK ");
    return 1;
  }
  else {
    fprintf(stderr,"Invariant doesn't hold in initial state!\n");
    if (PRINT) print_BDD(formula,stderr);
    if (DOT) dot_BDD(formula,stdout);
    if (COUNTER) print_example(formula,prFALSE,stderr);
    return 0;
  }
}

static void traverse_sums(ATermList sums, ATermList parameters, ATerm I) { 
  ATerm summand;
  ATerm b_a;
  ATerm formula;
  ATermList sum_a, f_a, g_a;
  char *a;
  int wrong=0,count=0;
  
  for( ; (!ATisEmpty(sums)) ; ) {
    summand = ATgetFirst(sums);
    sums = ATgetNext(sums);
    count++;
    if (!ATmatch(summand,"smd(<term>,<str>,<term>,i(<term>),<term>)",
		 &sum_a,&a,&f_a,&g_a,&b_a)) 
      ATerror("Expect summand %t",summand);
    
    // ATfprintf(stderr,"Action: %s\n",a);
    
    Declare_vars(sum_a);
    formula = prIMPLIES(prAND(b_a,I),substitute(g_a,parameters,I));

    if (GENERATE) {
      if (VERBOSE) ATfprintf(stderr,".");
      ATfprintf(stdout,",\n\n%t,\n",
		prettyprint_decls(ATconcat(parameters,sum_a)));
      ATfprintf(stdout,"%t",prettyprint(Simplify(formula)));
    }
    else {
      formula = Prove(formula);
      if (formula == prTRUE)
	ATfprintf(stderr,".");
      else {
	ATfprintf(stderr,"X");
	if (!ALL) {
	  sums=ATmakeList0();
	  ATfprintf(stderr,"\nNo! (summand %d violates invariant)\n",count);
	}
	if (PRINT) print_BDD(formula,stderr);
	if (DOT) dot_BDD(formula,stdout);
	if (COUNTER) print_example(formula,prFALSE,stderr);
	wrong=1;
      }
    }
  }
  if ((!GENERATE) && VERBOSE && !wrong)
    fprintf(stderr,"\nYes! invariant holds for all summands!");
  if (VERBOSE) fprintf(stderr,"\n");
}

static void check_invariance_conditions(ATerm LPO,ATerm I)
{
  ATerm init=NULL;
  ATermList parameters=NULL, sums=NULL; 

  if (!ATmatch(LPO,"initprocspec(<term>,<term>,<term>)",
	       &init,&parameters,&sums))
    ATerror("Expect LPO\n");
  
  Declare_vars(parameters);
  
  if (GENERATE) fprintf(stdout,"[");
  if (check_init((ATermList)init,parameters,I) || ALL)
    traverse_sums(sums,parameters,I);  
  if (GENERATE) fprintf(stdout,"]\n");
}

static void WarningHandler(const char *format, va_list args)
     {
     fprintf(stderr,"%s: ", MY_NAME);
     ATvfprintf(stderr, format, args);
     fprintf(stderr,"\n");
     }
     
static void ErrorHandler(const char *format, va_list args)
     {
     fprintf(stderr,"%s: ", MY_NAME);
     ATvfprintf(stderr, format, args);
     fprintf(stderr,"\n");
     exit(-1);
     }
     
static void ExitWrongArgument(int argc, char *argv[]) {
     if (argc>1) {
         ATwarning("Illegal option %s", argv[1]);
         help();
         exit(1);
         }
     }
     
int main(int argc, char *argv[])
{ ATerm bottomOfStack=NULL;
  ATerm LPO=NULL, I=NULL;
  FILE *fp;
  ATinit(argc,argv,&bottomOfStack);
  ATsetWarningHandler(WarningHandler);
  ATsetErrorHandler(ErrorHandler);
  ATprotect(&LPO);
  ATprotect(&I);

  parse_argument_line(&argc,&argv);
 
  MCRLsetArguments(&argc,&argv);
  if (!MCRLinitialize()) ATerror("Error with MCRLinitialize");

  LPO=MCRLgetProc();
  ProverSetArguments(&argc,&argv);
  ExitWrongArgument(argc, argv);
  ProverInitialize();

  MCRLdeclareVars(MCRLgetListOfPars());
  
  fp = fopen(Inv_file,"r");
  if (!fp) {
    perror(Inv_file);
    exit(1);
  }
  
  I = read_formula(fp);
  if (!I) 
    ATerror("Invariant file %s is wrong\n",Inv_file);
  
  check_invariance_conditions(LPO,I);
  
  return 0;
}
